Add config flow to DSMR integration and remove yaml support

RobBie1221 opened the pull request:
## Breaking change
This PR will deprecate YAML support for the DSMR integration and implement configuration through the UI.

## Proposed change
Config flow is added to the DSMR integration and support for yaml is removed.

Some other minor tweaks are performed, should_poll and force_update properties are set and an option is added to always have power data reported in Watt.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:


## Additional information
- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13899

## Checklist

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

RobBie1221 committed:
Add to DSMR integration parameters force_update and prefix

homeassistant commented:
Hi @RobBie1221, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#37464).

Once you do that we will be able to review and accept this pull request.

Thanks!

RobBie1221 committed:
Add prefix to gas entities

RobBie1221 committed:
Disable for DSMR sensor polling

frenck commented on homeassistant/components/dsmr/sensor.py:
```diff
@@ -47,6 +51,8 @@
         ),
         vol.Optional(CONF_RECONNECT_INTERVAL, default=30): int,
         vol.Optional(CONF_PRECISION, default=DEFAULT_PRECISION): vol.Coerce(int),
+        vol.Optional(CONF_FORCE_UPDATE, default=DEFAULT_FORCE_UPDATE): cv.boolean,
```
We no longer allow changes to the YAML configuration for platform integrations.

This integration needs to be moved into its own domain first. As this integration communicates with a device, it needs to be upgraded to use a configuration flow.

For more information see: https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md#decision
And: <https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision>

RobBie1221 replied:
I see that I have some reading to do. I guess that will make the PR a bit bigger. Would you propose to reject this pull request?

Out of curiosity, you may have seen that I'm quite new to home assistant development, is it just a matter of start typing when I would want to upgrade this integration or is it good practice to start a conversation somewhere?

frenck replied:
In the current state, it would be rejected, yes. I would propose to use this PR to bring the integration up to an upgraded state.

RobBie1221 replied:
Ok, so move it to it's own domain and make a configuration flow. Would I correctly conclude that the first one is a breaking change?

frenck replied:
Moving an integration configuration out of YAML into a config flow, is a breaking change. But before adding those sections, start building it first 😉

RobBie1221 replied:
Added config flow

RobBie1221 committed:
Move DSMR integration to its own domain

RobBie1221 committed:
Move DSMR integration to its own domain

RobBie1221 committed:
Add config flow for DSMR integration

RobBie1221 committed:
Fix strings for DSMR integration

RobBie1221 committed:
Rework DSMR tests based on config_flow implementation

RobBie1221 committed:
Add for DSMR integration config_flow tests and fix bugs

RobBie1221 committed:
Possibly fix #31604

RobBie1221 committed:
Add for DSMR integration parameter to always report in Watt, adjust tests

RobBie1221 committed:
Update tests for DSMR to increase test coverage

RobBie1221 committed:
Update tests for DSMR to increase test coverage

RobBie1221 committed:
Bugfix in DSMR test

RobBie1221 committed:
Add codeowner for DSMR integration

RobBie1221 committed:
Remove yaml support for DSMR

RobBie1221 committed:
Remove import config flow tests for DSMR

RobBie1221 committed:
Add already configured check to config flow

RobBie1221 committed:
Add tests for existing item config flow for DSMR

RobBie1221 committed:
Add some comments

RobBie1221 commented:
Dropped YAML support as suggested in the docs PR

RobBie1221 committed:
Some code quality improvements

RobBie1221 committed:
Make codecov happy

RobBie1221 committed:
Remove force_update as parameter

RobBie1221 committed:
Remove advanced options from config flow

RobBie1221 committed:
Implement option flow

RobBie1221 committed:
Adjust tests

RobBie1221 committed:
Add tests for option flow

RobBie1221 commented:
I changed the implemented a bit based on reviews I saw in other PR's:
- Removed force_update as parameter, always set to True (it's push integration with only sensor entities)
- Removed advanced options from config_flow to an options flow, not to bother users during adding with advanced properties.

RobBie1221 commented:
Not stale, waiting for review

jwnmulder commented on homeassistant/components/dsmr/const.py:
```diff
@@ -0,0 +1,27 @@
+"""Constants for the NEW_NAME integration."""
```
-> Constants for the DSMR integration.

RobBie1221 commented:
Closing this, PR is too big and I need to rework it anyway
